server/src/main/java/io/crate/expression/symbol/Symbol.java

```diff
@@ -59,6 +61,9 @@ public final Symbol cast(DataType<?> targetType) {
     public Symbol cast(DataType<?> targetType, boolean tryCast) {
         if (targetType.equals(valueType())) {
             return this;
+        } else if (ArrayType.unnest(targetType).equals(DataTypes.UNTYPED_OBJECT)
```

Comment:
so basically the cast for the `object` and `array(object)` was added coz the target type, the one that comes from the signature, has the empty inner types while the source data type has some inner types. There is no point to cast an object to an object type with empty inner types, so probably these semantic is not correct. I've added this condition in the `Symbol#cast`. @mfussenegger wdyt?

Reply:
Looks better to me to have it here :+1: